controllers:
  # joint states
  joint_state:
    type: joint_state_controller/JointStateController
    publish_rate: 200

  # separate position controllers for each actuator
  lf_shoulder_position:
    type: position_controllers/JointPositionController
    joint: lf_shoulder_joint
  lf_elbow_position:
    type: position_controllers/JointPositionController
    joint: lf_elbow_joint
  lf_wrist_position:
    type: position_controllers/JointPositionController
    joint: lf_wrist_joint
  rf_shoulder_position:
    type: position_controllers/JointPositionController
    joint: rf_shoulder_joint
  rf_elbow_position:
    type: position_controllers/JointPositionController
    joint: rf_elbow_joint
  rf_wrist_position:
    type: position_controllers/JointPositionController
    joint: rf_wrist_joint
  lm_shoulder_position:
    type: position_controllers/JointPositionController
    joint: lm_shoulder_joint
  lm_elbow_joint_position:
    type: position_controllers/JointPositionController
    joint: lm_elbow_joint
  lm_wrist_position:
    type: position_controllers/JointPositionController
    joint: lm_wrist_joint
  rm_shoulder_position:
    type: position_controllers/JointPositionController
    joint: rm_shoulder_joint
  rm_elbow_position:
    type: position_controllers/JointPositionController
    joint: rm_elbow_joint
  lb_shoulder_position:
    type: position_controllers/JointPositionController
    joint: lb_shoulder_joint
  lb_elbow_position:
    type: position_controllers/JointPositionController
    joint: lb_elbow_joint
  lb_wrist_position:
    type: position_controllers/JointPositionController
    joint: lb_wrist_joint
  rb_shoulder_position:
    type: position_controllers/JointPositionController
    joint: rb_shoulder_joint
  rb_elbow_position:
    type: position_controllers/JointPositionController
    joint: rb_elbow_joint
  rb_wrist_position:
    type: position_controllers/JointPositionController
    joint: rb_wrist_joint

  # position trajectory controller for all six legs
  group_position_trajectory:
    type: position_controllers/JointTrajectoryController
    joints:
      # left front leg
      - lf_shoulder_joint
      - lf_elbow_joint
      - lf_wrist_joint
      # right front leg
      - rf_shoulder_joint
      - rf_elbow_joint
      - rf_wrist_joint
      # left middle leg
      - lm_shoulder_joint
      - lm_elbow_joint
      - lm_wrist_joint
      # right middle leg
      - rm_shoulder_joint
      - rm_elbow_joint
      - rm_wrist_joint
      # left back leg
      - lb_shoulder_joint
      - lb_elbow_joint
      - lb_wrist_joint
      # right back leg
      - rb_shoulder_joint
      - rb_elbow_joint
      - rb_wrist_joint
    constraints:
      goal_time: 0.6
      stopped_velocity_tolerance: 0.05

  # cartesian controllers
  lf_cartesian_trajectory:
    type: "position_controllers/CartesianTrajectoryController"

    base: "base_link"
    tip: "lf_wrist_contact"

    joints:
      # left front leg
      - lf_shoulder_joint
      - lf_elbow_joint
      - lf_wrist_joint

  # custom arthropod controller
  arthropod_controller:
    type: "arthropod_controllers/ArthropodController"

    robot_base_frame: "base_link"
    odom_frame: "odom"
    robot_footprint_frame: "base_footprint"

    publish_rate: 50 # hz
    publish_cmd: true # default: false
    enable_odom_tf: true # default: true

    # legs
    legs:
      - "left_front"
      - "right_front"
      - "left_middle"
      - "right_middle"
      - "left_back"
      - "right_back"

    # leg controllers
    left_front:
      base: "base_link"
      tip: "lf_wrist_contact"

      joints:
        - lf_shoulder_joint
        - lf_elbow_joint
        - lf_wrist_joint

    right_front:
      base: "base_link"
      tip: "rf_wrist_contact"

      joints:
        - rf_shoulder_joint
        - rf_elbow_joint
        - rf_wrist_joint

    left_middle:
      base: "base_link"
      tip: "lm_wrist_contact"

      joints:
        - lm_shoulder_joint
        - lm_elbow_joint
        - lm_wrist_joint

    right_middle:
      base: "base_link"
      tip: "rm_wrist_contact"

      joints:
        - rm_shoulder_joint
        - rm_elbow_joint
        - rm_wrist_joint

    left_back:
      base: "base_link"
      tip: "lb_wrist_contact"

      joints:
        - lb_shoulder_joint
        - lb_elbow_joint
        - lb_wrist_joint

    right_back:
      base: "base_link"
      tip: "rb_wrist_contact"

      joints:
        - rb_shoulder_joint
        - rb_elbow_joint
        - rb_wrist_joint
